pkg/credentialprovider/azure/azure_credentials.go

```diff
@@ -186,6 +191,30 @@ func (a *acrProvider) Enabled() bool {
 }
 
 func (a *acrProvider) Provide(image string) credentialprovider.DockerConfig {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	// If the cache hasn't expired, return our cache
+	if time.Now().Before(a.expiration) {
+		if a.config != nil && a.config.UseManagedIdentityExtension {
+			loginServer := a.parseACRLoginServerFromImage(image)
+			if loginServer == "" {
+				return a.cacheDockerConfig
+			}
+
+			// check whether loginServer is in cache
+			for k := range a.cacheDockerConfig {
+				if strings.EqualFold(loginServer, k) {
+					return a.cacheDockerConfig
+				}
```

Comment:
We shouldn't add logic to this plugin that could not be expressed within the API of the externalized credential provider plugins, as that will prevent moving this credential provider out of tree.

Can you give a concrete example of what the `image` string will be, and what the `loginServer` string will be?

It seems like this plugin is fetching credentials based on the image and wants to cache them for any image using the same registry. Is that the case?

Reply:
@liggitt below is an example, with Azure Managed Identity (no password), the credential is based on `loginServer`, and different `loginServer` has different credentials.
 - `image`: `andyacr3.azurecr.io/nginx:v1`
 - `loginServer`: `andyacr3.azurecr.io`

AWS also uses same implementation today:
https://github.com/kubernetes/kubernetes/blob/4261200724cd0541b4ff02cc56be866b98142823/pkg/credentialprovider/aws/aws_credentials.go#L85-L98
I think externalized credential provider plugin should also supports such logic.

Anyway, this is now a critical bug in Azure side, we need to fix it ASAP, thanks for understanding, and will check how to bring this logic to externalized credential provider plugin later.

Reply:
Does the login server always match the registry portion of the image?

Reply:
depends on image itself, if it's Azure Container Registry image, e.g. `andyacr3.azurecr.io/nginx:v1`, then it would always match since we only cache Azure Container Registry login server.